cpp/src/arrow/flight/sql/driver.cc

```diff
@@ -0,0 +1,1877 @@
+        const std::optional<std::string>& cur_schema_name = schema_item.first;
+
+        if (cur_schema_name) {
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                   db_schema_name_builder->Append(*cur_schema_name));
+        } else {
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error, db_schema_name_builder->AppendNull());
+        }
+
+        if (depth == ADBC_OBJECT_DEPTH_ALL || depth >= ADBC_OBJECT_DEPTH_TABLES) {
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error, db_schema_tables_builder->Append());
+        } else {
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                   db_schema_tables_builder->AppendNull());
+          continue;
+        }
+
+        std::pair<int64_t, int64_t> schema_tables_range = schema_item.second;
+        for (int64_t i = schema_tables_range.first; i < schema_tables_range.second; i++) {
+          const std::string_view table_name = GetNotNull<StringType>(*table_data, 2, i);
+          const std::string_view table_type = GetNotNull<StringType>(*table_data, 3, i);
+          const std::string_view table_schema = GetNotNull<BinaryType>(*table_data, 4, i);
+
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                   db_schema_tables_items_builder->Append());
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                   table_name_builder->Append(table_name));
+          ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                   table_type_builder->Append(table_type));
+
+          if (depth == ADBC_OBJECT_DEPTH_COLUMNS) {
+            ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error, table_columns_builder->Append());
+
+            std::shared_ptr<Schema> schema;
+            io::BufferReader reader((Buffer(table_schema)));
+            ipc::DictionaryMemo memo;
+            ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                                     ipc::ReadSchema(&reader, &memo).Value(&schema));
+
+            int32_t ordinal_position = 1;
+            for (const std::shared_ptr<Field>& field : schema->fields()) {
+#ifdef ARROW_COMPUTE
+              if (column_name && std::strlen(column_name) > 0) {
+                Datum result;
+                arrow::compute::MatchSubstringOptions options(column_name);
+                ADBC_ARROW_RETURN_NOT_OK(
+                    INTERNAL, error,
+                    arrow::compute::CallFunction(
+                        "match_like", {Datum(MakeScalar(field->name()))}, &options)
+                        .Value(&result));
+                DCHECK_EQ(result.kind(), Datum::Kind::SCALAR);
+                if (!result.scalar_as<BooleanScalar>().value) continue;
+              }
+#else
+              if (column_name && std::strlen(column_name) > 0) {
+                return Status::NotImplemented(
+                    "Cannot filter on column name without ARROW_COMPUTE");
```

Comment:
Can you open an issue so that the required facilities (the `MatchSubstring` classes and the `MatchLike` initialization boilerplate) are moved somewhere to `arrow/util`?

Reply:
I thought about doing that here, but just wanted to get it working.

https://github.com/apache/arrow/issues/14864